storage/remote/queue_manager.go

```diff
@@ -1028,7 +1032,91 @@ func (s *shards) enqueue(ref chunks.HeadSeriesRef, data interface{}) bool {
 	}
 }
 
-func (s *shards) runShard(ctx context.Context, shardID int, queue chan interface{}) {
+type queue struct {
+	batch      []interface{}
+	batchQueue chan []interface{}
+
+	// Since we know there are a limited number of batches out, using a stack
+	// is easy and safe so a sync.Pool is not necessary.
+	batchPool [][]interface{}
+	// This mutex covers adding and removing batches from the batchPool.
+	poolMux sync.Mutex
+}
+
+func newQueue(batchSize, capacity int) *queue {
+	batches := capacity / batchSize
```

Comment:
If someone configures Capacity < SamplesPerSend, we get an unbuffered chan, right?
It's probably not desirable.

Reply:
enforcing capacity >= max samples per send sounds like a good idea to me

Reply:
Great, I will do that in a separate PR as it is orthogonal to this.